core/src/main/scala/cats/sequence/sequence.scala

```diff
@@ -17,85 +17,157 @@ trait Sequencer[L <: HList] extends Serializable {
+    type F[X] = F0[X]
+    type LOut = H :: HNil
+
+    def apply(singleton: F[H] :: HNil) = F.imap(singleton.head)(_ :: HNil)(_.head)
+  }
+}
 
+private[sequence] abstract class MkHConsSequencerLowPrio extends MkHNilSequencerLowPrio {
   type Aux[L <: HList, F0[_], LOut0] = Sequencer[L] {
     type F[X] = F0[X]
     type LOut = LOut0
   }
 
+  implicit def mkHConsSequencerForInvariant[F0[_], H, FT <: HList, T <: HList](
+                                                                                implicit tailSequencer: Aux[FT, F0, T], F: InvariantMonoidal[F0]
+                                                                              ): Aux[F0[H] :: FT, F0, H :: T] = new Sequencer[F0[H] :: FT] {
+    type F[X] = F0[X]
+    type LOut = H :: T
+
+    def apply(hl: F[H] :: FT) = F.imap(F.product(hl.head, tailSequencer(hl.tail)))(p => p._1 :: p._2)(hl => (hl.head, hl.tail))
+  }
+}
+
+private[sequence] abstract class MkHConsSequencer extends MkHConsSequencerLowPrio {
+
   implicit def mkHConsSequencer[F0[_], H, FT <: HList, T <: HList](
-    implicit tailSequencer: Aux[FT, F0, T], F: Apply[F0]
-  ): Aux[F0[H] :: FT, F0, H :: T] = new Sequencer[F0[H] :: FT] {
+                                                                    implicit tailSequencer: Aux[FT, F0, T], F: Apply[F0]
+                                                                  ): Aux[F0[H] :: FT, F0, H :: T] = new Sequencer[F0[H] :: FT] {
     type F[X] = F0[X]
     type LOut = H :: T
+
     def apply(hl: F[H] :: FT) = F.map2(hl.head, tailSequencer(hl.tail))(_ :: _)
   }
 }
 
-object Sequencer extends MkHConsSequencer {
 
+object Sequencer extends MkHConsSequencer {
   implicit def mkHNilSequencer[F0[_]](
-    implicit F: Applicative[F0]
-  ): Aux[HNil, F0, HNil] = new Sequencer[HNil] {
+                                       implicit F: Applicative[F0]
+                                     ): Aux[HNil, F0, HNil] = new Sequencer[HNil] {
     type F[X] = F0[X]
     type LOut = HNil
+
     def apply(nil: HNil) = F.pure(nil)
   }
 
   implicit def mkSingletonSequencer[F0[_], H](
-    implicit F: Functor[F0]
-  ): Aux[F0[H] :: HNil, F0, H :: HNil] = new Sequencer[F0[H] :: HNil] {
+                                               implicit F: Functor[F0]
+                                             ): Aux[F0[H] :: HNil, F0, H :: HNil] = new Sequencer[F0[H] :: HNil] {
     type F[X] = F0[X]
     type LOut = H :: HNil
+
     def apply(singleton: F[H] :: HNil) = F.map(singleton.head)(_ :: HNil)
   }
 }
 
 @implicitNotFound("cannot construct sequencer, make sure that every field value of you record ${L} is an Apply")
 trait RecordSequencer[L <: HList] extends Serializable {
   type Out
+
   def apply(record: L): Out
 }
 
-object RecordSequencer {
-  type Aux[L <: HList, Out0] = RecordSequencer[L] { type Out = Out0 }
+trait RecordSequencerInstanceLowPrio {
```

Comment:
This is to ensure we can evolve it later while preserving binary compatibility:
```suggestion
private[sequence] abstract class RecordSequencerInstanceLowPrio {
```